qiskit/transpiler/passes/optimization/optimize_1q_decomposition.py

```diff
@@ -69,31 +69,32 @@ def run(self, dag):
         if not self.basis:
             LOG.info("Skipping pass because no basis is set")
             return dag
-        decomposer = OneQubitEulerDecomposer(self.basis)
-        runs = dag.collect_runs(self.euler_basis_names[self.basis])
-        runs = _split_runs_on_parameters(runs)
-        for run in runs:
-            if len(run) <= 1:
-                params = run[0].op.params
-                # Remove single identity gates
-                if run[0].op.name in self.euler_basis_names[self.basis] and len(
-                        params) > 0 and np.array_equal(run[0].op.to_matrix(),
-                                                       np.eye(2)):
-                    dag.remove_op_node(run[0])
-                # Don't try to optimize a single 1q gate
-                continue
-            q = QuantumRegister(1, "q")
-            qc = QuantumCircuit(1)
-            for gate in run:
-                qc.append(gate.op, [q[0]], [])
-
-            operator = Operator(qc)
-            new_circ = decomposer(operator)
-            new_dag = circuit_to_dag(new_circ)
-            dag.substitute_node_with_dag(run[0], new_dag)
-            # Delete the other nodes in the run
-            for current_node in run[1:]:
-                dag.remove_op_node(current_node)
+        for basis in self.basis:
+            decomposer = OneQubitEulerDecomposer(basis)
+            runs = dag.collect_runs(self.euler_basis_names[basis])
```

Comment:
Do we need to collect runs over the 1q basis gates here? `Optimize1qGates` requires input gates to the optimization to be in `{'u1', 'u2', 'u3'}`, but the `OneQubitEulerDecomposer` accepts as input the 2x2 matrix from the `Operator`, so the particular input basis shouldn't matter. (Maybe this is why some of the optimizations for over-complete bases were longer than expected.)

Instead, maybe collect all runs of (non-parameterized) 1q gates, and decompose those across the available target bases, choosing whichever results in the shortest depth (or by some other metric).

Reply:
I had been treating things here the same way as we did in `Optimize1qGates` and were expecting things to be already be translated. But, yeah you're right we can just use any single qubit run and get it's matrix and it should work better. 

To do this though I think I'll have to add a new dagcircuit method to get all 1q gate runs. We might be able to do something with `collect_runs` and the basis list but the issue is trying to figure out the 1q gates from the basis strings is non-trivial because there could be custom gates in there.

Reply:
Ok, I took care of this. I added the 1q run collection method in https://github.com/Qiskit/qiskit-terra/pull/5431/commits/9271b32a5aef65099d8d4b7c599e46f42af0687a and switched this pass to use that in https://github.com/Qiskit/qiskit-terra/pull/5431/commits/1d16fd8551fcaa442c77e18d5db77b9c192ff01a

Reply:
Related to this I also just pushed up: https://github.com/Qiskit/retworkx/pull/206 which adds a retworkx native function for the nested loops used to search for runs in the graph. Once that is merged and released we can use that for both `collect_runs` and `collect_1q_runs`.